File: server/src/routes/corpus.ts
// /api/corpus/{select,selectOne}
//
// Mirror the CorpusAdapter shape — these endpoints are POST (not GET) because
// SQL strings can exceed URL length limits and may contain characters that
// would need escaping. The frontend's web adapter (Wave 3b) will call these
// with the same `{ sql, params }` body shape that the Tauri plugin accepts.

import { Router } from "express";
import type { CorpusHandle } from "../corpus";
import { QueryError } from "../corpus";

interface QueryBody {
  sql?: unknown;
  params?: unknown;
}

function normalizeParams(raw: unknown): unknown[] {
  if (raw === undefined || raw === null) return [];
  if (!Array.isArray(raw)) {
    throw new QueryError("params must be an array", 400);
  }
  return raw;
}

export function corpusRouter(corpus: CorpusHandle): Router {
  const router = Router();

  router.post("/select", (req, res) => {
    const { sql, params } = req.body as QueryBody;
    if (typeof sql !== "string") {
      res.status(400).json({ error: "sql must be a string" });
      return;
    }
    try {
      const rows = corpus.select(sql, normalizeParams(params));
      res.json({ rows });
    } catch (err) {
      handleErr(res, err);
    }
  });

  router.post("/selectOne", (req, res) => {
    const { sql, params } = req.body as QueryBody;
    if (typeof sql !== "string") {
      res.status(400).json({ error: "sql must be a string" });
      return;
    }
    try {
      const row = corpus.selectOne(sql, normalizeParams(params));
      res.json({ row });
    } catch (err) {
      handleErr(res, err);
    }
  });

  return router;
}

function handleErr(res: import("express").Response, err: unknown): void {
  if (err instanceof QueryError) {
    res.status(err.status).json({ error: err.message });
    return;
  }
  const message = err instanceof Error ? err.message : String(err);
  res.status(400).json({ error: message });
}

File: server-py/app/routes/corpus.py
"""/api/corpus/{select,selectOne} — Postgres backend (phase 2).

The request/response shape is unchanged from phase 1: clients post
``{sql, params}`` and get back ``{rows}`` or ``{row}``. FTS5 idioms
(``verse_fts MATCH``, ``snippet(...)``, ``ORDER BY rank``) in the incoming
SQL are translated to Postgres tsvector equivalents by ``app.db.rewrite_fts``
before execution.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..db import QueryError, get_pool, select, select_one


def _normalize_params(raw: Any) -> list[Any]:
    if raw is None:
        return []
    if not isinstance(raw, list):
        raise QueryError("params must be an array", 400)
    return raw


def _err(message: str, status: int = 400) -> JSONResponse:
    return JSONResponse({"error": message}, status_code=status)


def corpus_router() -> APIRouter:
    router = APIRouter()

    @router.post("/select")
    async def select_endpoint(req: Request) -> JSONResponse:
        try:
            body = await req.json()
        except Exception:
            return _err("invalid JSON body")
        if not isinstance(body, dict):
            return _err("body must be a JSON object")
        sql = body.get("sql")
        if not isinstance(sql, str):
            return _err("sql must be a string")
        try:
            params = _normalize_params(body.get("params"))
        except QueryError as err:
            return _err(str(err), err.status)
        pool = await get_pool(req.app.state)
        try:
            rows = await select(pool, sql, params)
            return JSONResponse({"rows": rows})
        except QueryError as err:
            return _err(str(err), err.status)
        except Exception as err:  # noqa: BLE001
            return _err(str(err), 400)

    @router.post("/selectOne")
    async def select_one_endpoint(req: Request) -> JSONResponse:
        try:
            body = await req.json()
        except Exception:
            return _err("invalid JSON body")
        if not isinstance(body, dict):
            return _err("body must be a JSON object")
        sql = body.get("sql")
        if not isinstance(sql, str):
            return _err("sql must be a string")
        try:
            params = _normalize_params(body.get("params"))
        except QueryError as err:
            return _err(str(err), err.status)
        pool = await get_pool(req.app.state)
        try:
            row = await select_one(pool, sql, params)
            return JSONResponse({"row": row})
        except QueryError as err:
            return _err(str(err), err.status)
        except Exception as err:  # noqa: BLE001
            return _err(str(err), 400)

    return router
